util/topsql/reporter/client.go

```diff
@@ -20,45 +20,120 @@ import (
 	"sync"
 	"time"
 
+	"github.com/pingcap/tidb/config"
+	"github.com/pingcap/tidb/util"
 	"github.com/pingcap/tidb/util/logutil"
 	"github.com/pingcap/tipb/go-tipb"
 	"go.uber.org/zap"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/backoff"
 )
 
-// ReportClient send data to the target server.
+// ReportClient sends data to the target server.
 type ReportClient interface {
-	Send(ctx context.Context, addr string, data reportData) error
+	Send(data reportData, timeout time.Duration)
+
+	// IsPending indicates that ReportClient is not expecting to receive records for now and may resume in the future.
+	IsPending() bool
+
+	// IsDown indicates that the client has been down and can be cleared.
+	// Note that: once a ReportClient is down, it cannot go back to be up.
+	IsDown() bool
+
 	Close()
 }
 
-// GRPCReportClient reports data to grpc servers.
-type GRPCReportClient struct {
+// ReportClientRegistry is used to receive ReportClient registrations.
+type ReportClientRegistry struct {
+	sync.Mutex
+	newClients []ReportClient
+}
+
+// NewReportClientRegistry creates a new ReportClientRegistry.
+func NewReportClientRegistry() *ReportClientRegistry {
+	return &ReportClientRegistry{}
+}
+
+func (r *ReportClientRegistry) visitAndClear(visit func(client ReportClient)) {
+	r.Lock()
+	defer r.Unlock()
+
+	for i := range r.newClients {
+		visit(r.newClients[i])
+	}
+	r.newClients = r.newClients[:0]
+}
+
+func (r *ReportClientRegistry) register(client ReportClient) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.newClients = append(r.newClients, client)
+}
+
+// ConfigAssignedRemoteClient reports data to grpc servers.
+type ConfigAssignedRemoteClient struct {
 	curRPCAddr string
 	conn       *grpc.ClientConn
+	sendTaskCh chan sendTask
 	// calling decodePlan this can take a while, so should not block critical paths
 	decodePlan planBinaryDecodeFunc
 }
 
-// NewGRPCReportClient returns a new GRPCReportClient
-func NewGRPCReportClient(decodePlan planBinaryDecodeFunc) *GRPCReportClient {
-	return &GRPCReportClient{
+type sendTask struct {
+	data    reportData
+	timeout time.Duration
+}
+
+// NewGRPCReportClient returns a new ConfigAssignedRemoteClient
+func NewGRPCReportClient(decodePlan planBinaryDecodeFunc) *ConfigAssignedRemoteClient {
+	client := &ConfigAssignedRemoteClient{
 		decodePlan: decodePlan,
+		sendTaskCh: make(chan sendTask),
+	}
+
+	go util.WithRecovery(client.run, nil)
+	return client
+}
+
+func (r *ConfigAssignedRemoteClient) run() {
+	for task := range r.sendTaskCh {
+		targetRPCAddr := config.GetGlobalConfig().TopSQL.ReceiverAddress
+		if targetRPCAddr == "" {
+			continue
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), task.timeout)
+		start := time.Now()
```

Comment:
Maybe using https://github.com/prometheus/client_golang/blob/v1.11.0/prometheus/timer.go#L33 (which specifically indicate that this timer is for recording histogram seconds) is more idiomatic.

Reply:
Don't know which observer to observe the result in advance, but
```go
func NewTimer(o Observer) *Timer 
```
needs to know.